README.md
## Getting started

- [x] webvtt suporrt
- [x] srt suporrt
- [x] multiple line text
- [x] extract html

![](https://i.ibb.co/BGcxskh/image.png)

```dart
import 'dart:convert' show utf8;

import 'package:chewie/chewie.dart';
import 'package:flutter/material.dart';
import 'package:flutter_subtitle/flutter_subtitle.dart' hide Subtitle;
import 'package:video_player/video_player.dart';
import 'package:http/http.dart' as http;

void main() => runApp(const VideoApp());

class VideoApp extends StatefulWidget {
  const VideoApp({Key? key}) : super(key: key);

  @override
  State createState() => _VideoAppState();
}

class _VideoAppState extends State<VideoApp> {
  ChewieController? _chewieController;
  late SubtitleController _subtitleController;

  @override
  void initState() {
    getWebVtt();
    super.initState();
  }

  getWebVtt() async {
    // https://cc.zorores.com/20/2e/202eaab6dff289a5976399077449654e/eng-2.vtt
    // https://ohplayer.netlify.app/%E5%90%9B%E3%81%AE%E5%90%8D%E3%81%AF.srt

    final body = utf8.decode((await http.get(Uri.parse(
            'https://cc.zorores.com/20/2e/202eaab6dff289a5976399077449654e/eng-2.vtt')))
        .bodyBytes);

    _subtitleController =
        SubtitleController.string(body, format: SubtitleFormat.webvtt);
    final controller = VideoPlayerController.network(
      'https://ohplayer.netlify.app/%E5%90%9B%E3%81%AE%E5%90%8D%E3%81%AF.mp4',
    );

    await controller.initialize();

    controller.addListener(() {
      setState(() {});
    });

    _chewieController = ChewieController(
      videoPlayerController: controller,
      autoPlay: true,
      looping: true,
      subtitleBuilder: (context, subtitle) {
        return IgnorePointer(
          child: SubtitleView(
            text: subtitle,
            subtitleStyle: SubtitleStyle(
              fontSize: _chewieController!.isFullScreen ? 20 : 16,
            ),
          ),
        );
      },
    );

    //// Chewie Subtitle
    _chewieController!.setSubtitle(
      _subtitleController.subtitles
          .map(
            (e) => Subtitle(
              index: e.number,
              start: Duration(milliseconds: e.start),
              end: Duration(milliseconds: e.end),
              text: e.text,
            ),
          )
          .toList(),
    );

    //// Caption
    // final webVTTCaptionFile = _subtitleController.subtitles
    //     .map((e) => Caption(
    //           number: e.number,
    //           start: Duration(milliseconds: e.start),
    //           end: Duration(milliseconds: e.end),
    //           text: e.text,
    //         ))
    //     .toList();

    // controller.setClosedCaptionFile(
    //     Future.value(MyWebVTTCaptionFile(webVTTCaptionFile)));

    controller.setClosedCaptionFile(Future.value(WebVTTCaptionFile(body)));
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Video Demo',
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.start,
          children: [
            const SizedBox(height: 200),
            _chewieController != null
                ? AspectRatio(
                    aspectRatio: 16 / 9,
                    child: Chewie(controller: _chewieController!),
                  )
                : const CircularProgressIndicator(),
            const SizedBox(height: 20),
            const Text('SubtitleControllView'),
            if (_chewieController != null)
              SubtitleControllView(
                subtitleController: _subtitleController,
                inMilliseconds: _chewieController!
                    .videoPlayerController.value.position.inMilliseconds,
              ),
            const SizedBox(height: 20),
            const Text('ClosedCaption'),
            if (_chewieController != null)
              ClosedCaption(
                text:
                    _chewieController!.videoPlayerController.value.caption.text,
                textStyle: const TextStyle(
                  fontSize: 16,
                  color: Colors.white,
                ),
              ),
          ],
        ),
      ),
    );
  }

  @override
  void dispose() {
    super.dispose();
    _chewieController?.dispose();
  }
}

class MyWebVTTCaptionFile extends ClosedCaptionFile {
  MyWebVTTCaptionFile(this.captions);

  @override
  List<Caption> captions = [];
}
```


CHANGELOG.md
## 0.0.4

- Improve newline character handling in WebVTT parsing.

## 0.0.3

- suporrt webvtt and srt

## 0.0.2

- initial release.

## 0.0.1

- initial release.


test/srt_test.dart
// ignore_for_file: unused_element

import 'package:flutter_subtitle/flutter_subtitle.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('Parses SubRip file', () {
    final parsedFile = SubtitleController.string(
      _validSubRip,
      format: SubtitleFormat.srt,
    );

    expect(parsedFile.subtitles.length, 4);

    final Subtitle firstSubtitle = parsedFile.subtitles.first;
    expect(firstSubtitle.number, 1);
    expect(firstSubtitle.start, const Duration(seconds: 6).inMilliseconds);
    expect(firstSubtitle.end,
        const Duration(seconds: 12, milliseconds: 74).inMilliseconds);
    expect(firstSubtitle.text, 'This is a test file');

    final Subtitle secondSubtitle = parsedFile.subtitles[1];
    expect(secondSubtitle.number, 2);
    expect(
      secondSubtitle.start,
      const Duration(minutes: 1, seconds: 54, milliseconds: 724).inMilliseconds,
    );
    expect(
      secondSubtitle.end,
      const Duration(minutes: 1, seconds: 56, milliseconds: 760).inMilliseconds,
    );
    expect(secondSubtitle.text, '- Hello.\n- Yes?');

    final Subtitle thirdSubtitle = parsedFile.subtitles[2];
    expect(thirdSubtitle.number, 3);
    expect(
      thirdSubtitle.start,
      const Duration(minutes: 1, seconds: 56, milliseconds: 884).inMilliseconds,
    );
    expect(
      thirdSubtitle.end,
      const Duration(minutes: 1, seconds: 58, milliseconds: 954).inMilliseconds,
    );
    expect(
      thirdSubtitle.text,
      'These are more test lines\nYes, these are more test lines.',
    );

    final Subtitle fourthSubtitle = parsedFile.subtitles[3];
    expect(fourthSubtitle.number, 4);
    expect(
      fourthSubtitle.start,
      const Duration(hours: 1, minutes: 1, seconds: 59, milliseconds: 84)
          .inMilliseconds,
    );
    expect(
      fourthSubtitle.end,
      const Duration(hours: 1, minutes: 2, seconds: 1, milliseconds: 552)
          .inMilliseconds,
    );
    expect(
      fourthSubtitle.text,
      "- [ Machinery Beeping ]\n- I'm not sure what that was,",
    );
  });

  test('Parses SubRip file with malformed input', () {
    final parsedFile =
        SubtitleController.string(_malformedSubRip, format: SubtitleFormat.srt);

    expect(parsedFile.subtitles.length, 1);

    // final Subtitle firstSubtitle = parsedFile.subtitles.single;
    // expect(firstSubtitle.number, 2);
    // expect(firstSubtitle.start, const Duration(seconds: 15).inMilliseconds);
    // expect(firstSubtitle.end,
    //     const Duration(seconds: 17, milliseconds: 74).inMilliseconds);
    // expect(firstSubtitle.text, 'This one is valid');
  });
}

const String _validShortSubRip = '''
1
00:00:06,000 --> 00:00:12,074
This is a test file

2
00:01:54,724 --> 00:01:56,760
- Hello.
- Yes?
''';

const String _validSubRip = '''
1
00:00:06,000 --> 00:00:12,074
This is a test file

2
00:01:54,724 --> 00:01:56,760
- Hello.
- Yes?

3
00:01:56,884 --> 00:01:58,954
These are more test lines
Yes, these are more test lines.

4
01:01:59,084 --> 01:02:01,552
- [ Machinery Beeping ]
- I'm not sure what that was,

''';

const String _malformedSubRip = '''
1
00:00:06,000--> 00:00:12,074
This one should be ignored because the
arrow needs a space.

2
00:00:15,000 --> 00:00:17,074
This one is valid

3
00:01:54,724 --> 00:01:6,760
This one should be ignored because the
ned time is missing a digit.
''';


test/utils_test.dart
// ignore_for_file: constant_identifier_names, avoid_print, unused_import, unused_element

import 'package:flutter_subtitle/flutter_subtitle.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('test utils', () {
    // final lineSections = readSubtitleFile(_valid_vtt_with_metadata);
    // final regSections = matchSubtitleStrings(_valid_vtt_with_metadata);

    // print(lineSections);
    // print(regSections);
  });
}

const String _malformedSubRip = '''
1
00:00:06,000--> 00:00:12,074
This one should be ignored because the
arrow needs a space.

2
00:00:15,000 --> 00:00:17,074
This one is valid

3
00:01:54,724 --> 00:01:6,760
This one should be ignored because the
ned time is missing a digit.
''';

const String _valid_vtt_with_metadata = '''
WEBVTT Kind: subtitles; Language: en

REGION
id:bill
width:40%
lines:3
regionanchor:100%,100%
viewportanchor:90%,90%
scroll:up

NOTE
This file was written by Jill. I hope
you enjoy reading it. Some things to
bear in mind:
- I was lip-reading, so the cues may
not be 100% accurate
- I didn’t pay too close attention to
when the cues should start or end.

1
00:01.000 --> 00:02.500
<v Roger Bingham>We are in New York City''';


test/webvtt_test.dart
// ignore_for_file: constant_identifier_names

import 'package:flutter_subtitle/flutter_subtitle.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Parse VTT file', () {
    SubtitleController parsedFile;

    test('with Metadata', () {
      parsedFile = SubtitleController.string(
        _valid_vtt_with_metadata,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles.length, 1);

      expect(parsedFile.subtitles[0].start,
          const Duration(seconds: 1).inMilliseconds);
      expect(parsedFile.subtitles[0].end,
          const Duration(seconds: 2, milliseconds: 500).inMilliseconds);
      expect(parsedFile.subtitles[0].text, 'We are in New York City');
    });

    test('with Multiline', () {
      parsedFile = SubtitleController.string(
        _valid_vtt_with_multiline,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles.length, 1);

      expect(parsedFile.subtitles[0].start,
          const Duration(seconds: 2, milliseconds: 800).inMilliseconds);
      expect(parsedFile.subtitles[0].end,
          const Duration(seconds: 3, milliseconds: 283).inMilliseconds);
      expect(parsedFile.subtitles[0].text,
          '— It will perforate your stomach.\n— You could die.');
    });

    test('with various line terminators', () {
      parsedFile = SubtitleController.string(
        _valid_vtt_with_newlines,
        format: SubtitleFormat.webvtt,
      );

      final expectations = [
        Subtitle(
          number: 1,
          start: const Duration(seconds: 0).inMilliseconds,
          end: const Duration(seconds: 1).inMilliseconds,
          text: 'Line feed subtitle.',
        ),
        Subtitle(
          number: 2,
          start: const Duration(seconds: 2).inMilliseconds,
          end: const Duration(seconds: 3).inMilliseconds,
          text: 'Carriage return subtitle.',
        ),
        Subtitle(
          number: 3,
          start: const Duration(seconds: 4).inMilliseconds,
          end: const Duration(seconds: 5).inMilliseconds,
          text: 'CR/LF subtitle.',
        ),
        Subtitle(
          number: 4,
          start: const Duration(seconds: 6).inMilliseconds,
          end: const Duration(seconds: 7).inMilliseconds,
          text: 'Chaos subtitle.',
        ),
      ];

      expect(parsedFile.subtitles.length, expectations.length);
      for (var i = 0; i < expectations.length; i++) {
        expect(parsedFile.subtitles[i].toString(), expectations[i].toString());
      }
    });

    test('with styles tags', () {
      parsedFile = SubtitleController.string(
        _valid_vtt_with_styles,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles.length, 3);

      expect(parsedFile.subtitles[0].start,
          const Duration(seconds: 5, milliseconds: 200).inMilliseconds);
      expect(parsedFile.subtitles[0].end,
          const Duration(seconds: 6).inMilliseconds);
      expect(parsedFile.subtitles[0].text,
          "You know I'm so excited my glasses are falling off here.");
    });

    test('with subtitling features', () {
      parsedFile = SubtitleController.string(
        _valid_vtt_with_subtitling_features,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles.length, 3);

      expect(parsedFile.subtitles[0].number, 1);
      expect(parsedFile.subtitles.last.start,
          const Duration(seconds: 4).inMilliseconds);
      expect(parsedFile.subtitles.last.end,
          const Duration(seconds: 5).inMilliseconds);
      expect(parsedFile.subtitles.last.text, 'Transcrit par Célestes™');
    });

    test('with [hours]:[minutes]:[seconds].[milliseconds].', () {
      parsedFile = SubtitleController.string(
        _valid_vtt_with_hours,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles.length, 1);

      expect(parsedFile.subtitles[0].number, 1);
      expect(parsedFile.subtitles.last.start,
          const Duration(seconds: 1).inMilliseconds);
      expect(parsedFile.subtitles.last.end,
          const Duration(seconds: 2).inMilliseconds);
      expect(parsedFile.subtitles.last.text, 'This is a test.');
    });

    test('with [minutes]:[seconds].[milliseconds].', () {
      parsedFile = SubtitleController.string(
        _valid_vtt_without_hours,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles.length, 1);

      expect(parsedFile.subtitles[0].number, 1);
      expect(parsedFile.subtitles.last.start,
          const Duration(seconds: 3).inMilliseconds);
      expect(parsedFile.subtitles.last.end,
          const Duration(seconds: 4).inMilliseconds);
      expect(parsedFile.subtitles.last.text, 'This is a test.');
    });

    test('with invalid seconds format returns empty subtitles.', () {
      parsedFile = SubtitleController.string(
        _invalid_seconds,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles, isEmpty);
    });

    test('with invalid minutes format returns empty subtitles.', () {
      parsedFile = SubtitleController.string(
        _invalid_minutes,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles, isEmpty);
    });

    test('with invalid hours format returns empty subtitles.', () {
      parsedFile = SubtitleController.string(
        _invalid_hours,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles, isEmpty);
    });

    test('with invalid component length returns empty subtitles.', () {
      parsedFile = SubtitleController.string(
        _time_component_too_long,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles, isEmpty);

      parsedFile = SubtitleController.string(
        _time_component_too_short,
        format: SubtitleFormat.webvtt,
      );
      expect(parsedFile.subtitles, isEmpty);
    });
  });

  test('Parses VTT file with malformed input.', () {
    final SubtitleController parsedFile = SubtitleController.string(
      _malformedVTT,
      format: SubtitleFormat.webvtt,
    );

    expect(parsedFile.subtitles.length, 1);

    final Subtitle firstCaption = parsedFile.subtitles.single;
    expect(firstCaption.number, 1);
    expect(firstCaption.start, const Duration(seconds: 13).inMilliseconds);
    expect(firstCaption.end, const Duration(seconds: 16).inMilliseconds);
    expect(firstCaption.text, 'Valid');
  });
}

/// See https://www.w3.org/TR/webvtt1/#introduction-comments
const String _valid_vtt_with_metadata = '''
WEBVTT Kind: subtitles; Language: en

REGION
id:bill
width:40%
lines:3
regionanchor:100%,100%
viewportanchor:90%,90%
scroll:up

NOTE
This file was written by Jill. I hope
you enjoy reading it. Some things to
bear in mind:
- I was lip-reading, so the cues may
not be 100% accurate
- I didn’t pay too close attention to
when the cues should start or end.

1
00:01.000 --> 00:02.500
<v Roger Bingham>We are in New York City''';

/// See https://www.w3.org/TR/webvtt1/#introduction-multiple-lines
const String _valid_vtt_with_multiline = '''
WEBVTT

2
00:02.800 --> 00:03.283
— It will perforate your stomach.
— You could die.

''';

/// https://www.w3.org/TR/webvtt1/#file-structure
const String _valid_vtt_with_newlines = '''WEBVTT
\n00:00:00.000 --> 00:00:01.000\nLine feed subtitle.\n
\r00:00:02.000 --> 00:00:03.000\rCarriage return subtitle.\r
\r\n00:00:04.000 --> 00:00:05.000\r\nCR/LF subtitle.\r\n
\n00:00:06.000 --> 00:00:07.000\r\nChaos subtitle.\r
''';

/// See https://www.w3.org/TR/webvtt1/#styling
const String _valid_vtt_with_styles = '''
WEBVTT

00:05.200 --> 00:06.000 align:start size:50%
<v Roger Bingham><i>You know I'm so excited my glasses are falling off here.</i>

00:00:06.050 --> 00:00:06.150
<v Roger Bingham><i>I have a different time!</i>

00:06.200 --> 00:06.900
<c.yellow.bg_blue>This is yellow text on a blue background</c>

''';

//See https://www.w3.org/TR/webvtt1/#introduction-other-features
const String _valid_vtt_with_subtitling_features = '''
WEBVTT

test
00:00.000 --> 00:02.000
This is a test.

Slide 1
00:00:00.000 --> 00:00:10.700
Title Slide

crédit de transcription
00:04.000 --> 00:05.000
Transcrit par Célestes™

''';

/// With format [hours]:[minutes]:[seconds].[milliseconds]
const String _valid_vtt_with_hours = '''
WEBVTT

test
00:00:01.000 --> 00:00:02.000
This is a test.

''';

/// Invalid seconds format.
const String _invalid_seconds = '''
WEBVTT

60:00:000.000 --> 60:02:000.000
This is a test.

''';

/// Invalid minutes format.
const String _invalid_minutes = '''
WEBVTT

60:60:00.000 --> 60:70:00.000
This is a test.

''';

/// Invalid hours format.
const String _invalid_hours = '''
WEBVTT

5:00:00.000 --> 5:02:00.000
This is a test.

''';

/// Invalid seconds format.
const String _time_component_too_long = '''
WEBVTT

60:00:00:00.000 --> 60:02:00:00.000
This is a test.

''';

/// Invalid seconds format.
const String _time_component_too_short = '''
WEBVTT

60:00.000 --> 60:02.000
This is a test.

''';

/// With format [minutes]:[seconds].[milliseconds]
const String _valid_vtt_without_hours = '''
WEBVTT

00:03.000 --> 00:04.000
This is a test.

''';

const String _malformedVTT = '''

WEBVTT Kind: subtitles; Language: en

00:09.000--> 00:11.430
<Test>This one should be ignored because the arrow needs a space.

00:13.000 --> 00:16.000
<Test>Valid

00:16.000 --> 00:8.000
<Test>This one should be ignored because the time is missing a digit.

''';


lib/src/subtitle_view.dart
import 'package:flutter/material.dart';
import '../flutter_subtitle.dart';

class SubtitleControllView extends StatelessWidget {
  final SubtitleController subtitleController;
  final int inMilliseconds;

  final Color? backgroundColor;
  final SubtitleStyle subtitleStyle;
  final EdgeInsets padding;

  const SubtitleControllView({
    Key? key,
    required this.inMilliseconds,
    required this.subtitleController,
    this.padding = const EdgeInsets.all(4),
    this.backgroundColor = const Color.fromRGBO(0, 0, 0, 0.6),
    this.subtitleStyle = const SubtitleStyle(),
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final text = subtitleController.textFromMilliseconds(
      inMilliseconds,
      subtitleController.subtitles,
    );

    if (text == '') return const SizedBox.shrink();

    return SubtitleView(
      text: text,
      padding: padding,
      subtitleStyle: subtitleStyle,
      backgroundColor: backgroundColor,
    );
  }
}

class SubtitleView extends StatelessWidget {
  final Color? backgroundColor;
  final SubtitleStyle subtitleStyle;
  final EdgeInsets padding;
  final String text;

  const SubtitleView({
    Key? key,
    required this.text,
    this.backgroundColor = const Color.fromRGBO(0, 0, 0, 0.6),
    this.subtitleStyle = const SubtitleStyle(),
    this.padding = const EdgeInsets.symmetric(horizontal: 4.0, vertical: 2),
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Align(
      alignment: Alignment.bottomCenter,
      child: DecoratedBox(
        decoration: BoxDecoration(
          color: backgroundColor,
          borderRadius: BorderRadius.circular(2.0),
        ),
        child: Padding(
          padding: padding,
          child: Stack(
            children: [
              if (subtitleStyle.bordered)
                Text(
                  text,
                  textAlign: TextAlign.center,
                  style: TextStyle(
                    fontSize: subtitleStyle.fontSize,
                    foreground: Paint()
                      ..style = subtitleStyle.borderStyle.style
                      ..strokeWidth = subtitleStyle.borderStyle.strokeWidth
                      ..color = subtitleStyle.borderStyle.color,
                  ),
                ),
              Text(
                text,
                textAlign: TextAlign.center,
                style: TextStyle(
                  color: subtitleStyle.textColor,
                  fontSize: subtitleStyle.fontSize,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

const _defaultStrokeWidth = 2.0;

class SubtitleBorderStyle {
  const SubtitleBorderStyle({
    this.strokeWidth = _defaultStrokeWidth,
    this.style = PaintingStyle.stroke,
    this.color = Colors.black,
  });

  final double strokeWidth;
  final PaintingStyle style;
  final Color color;
}

const _defaultFontSize = 16.0;

class SubtitleStyle {
  const SubtitleStyle({
    this.bordered = false,
    this.borderStyle = const SubtitleBorderStyle(),
    this.fontSize = _defaultFontSize,
    this.textColor = Colors.white,
  });
  final bool bordered;
  final SubtitleBorderStyle borderStyle;
  final double fontSize;
  final Color textColor;
}


lib/src/subtitle_utils.dart
import 'dart:convert';

import '../flutter_subtitle.dart';

List<Subtitle> parseSubtitleString(String data, SubtitleFormat format) {
  switch (format) {
    case SubtitleFormat.webvtt:
      return parseFromWebVTTString(data);
    case SubtitleFormat.srt:
      return parseFromSubRipString(data);
    default:
      return [];
  }
}

// video player
List<List<String>> readSubtitleFile(String file) {
  final List<String> lines = LineSplitter.split(file).toList();

  final List<List<String>> subtitleStrings = <List<String>>[];
  List<String> currentCaption = <String>[];
  int lineIndex = 0;
  for (final String line in lines) {
    final bool isLineBlank = line.trim().isEmpty;
    if (!isLineBlank) {
      currentCaption.add(line);
    }

    if (isLineBlank || lineIndex == lines.length - 1) {
      subtitleStrings.add(currentCaption);
      currentCaption = <String>[];
    }

    lineIndex += 1;
  }

  return subtitleStrings;
}

List<Subtitle> parseFromWebVTTString(String data) {
  final subtitles = <Subtitle>[];
  final captions = RegExp(_webVTTSegment).allMatches(data);

  int subtitleNumber = 1;

  for (final caption in captions) {
    final start = _timestampToMillisecond(caption.groups([2, 3, 4, 5]));
    final end = _timestampToMillisecond(caption.groups([6, 7, 8, 9]));
    final text = caption.group(11);
    if (start == null || end == null) {
      continue;
    }

    subtitles.add(Subtitle(
      number: int.tryParse(caption.group(1) ?? '') ?? subtitleNumber,
      start: start.inMilliseconds,
      end: end.inMilliseconds,
      text: extractTextFromHtml(text ?? ''),
    ));
    subtitleNumber++;
  }

  return subtitles;
}

Duration? _timestampToMillisecond(List<String?> segments) {
  final hours = int.parse(segments[0]?.split(':')[0] ?? '0');
  final minutes = int.parse(segments[1]?.split(':')[0] ?? '0');
  final seconds = int.parse(segments[2] ?? '');
  final milliseconds = int.parse(segments[3] ?? '0');
  if (minutes > 59 || seconds > 59) {
    return null;
  }

  return Duration(
    hours: hours,
    minutes: minutes,
    seconds: seconds,
    milliseconds: milliseconds,
  );
}

List<Subtitle> parseFromSubRipString(String data) {
  final List<Subtitle> subtitles = <Subtitle>[];
  final subtitleStrings = readSubtitleFile(data);

  for (final List<String> subtitleLine in subtitleStrings) {
    final range = rangefromSubRipString(subtitleLine[1]);

    if (range == null) {
      continue;
    }

    subtitles.add(Subtitle(
      number: int.parse(subtitleLine[0]),
      start: range[0].inMilliseconds,
      end: range[1].inMilliseconds,
      text: extractTextFromHtml(subtitleLine.sublist(2).join('\n')),
    ));
  }

  return subtitles;
}

// Parses a time stamp in an SubRip file into a Duration.
// For example:
//
// _parseSubRipTimestamp('00:01:59,084')
// returns
// Duration(hours: 0, minutes: 1, seconds: 59, milliseconds: 084)
Duration? parseSubRipTimestamp(String timestampString) {
  if (!RegExp(_subRipTimeStamp).hasMatch(timestampString)) {
    return null;
  }

  final List<String> commaSections = timestampString.split(',');
  final List<String> hoursMinutesSeconds = commaSections[0].split(':');

  final int hours = int.parse(hoursMinutesSeconds[0]);
  final int minutes = int.parse(hoursMinutesSeconds[1]);
  final int seconds = int.parse(hoursMinutesSeconds[2]);
  final int milliseconds = int.parse(commaSections[1]);

  return Duration(
    hours: hours,
    minutes: minutes,
    seconds: seconds,
    milliseconds: milliseconds,
  );
}

// 00:00:06,000 --> 00:00:12,074
List<Duration>? rangefromSubRipString(String line) {
  final RegExp format = RegExp(_subRipTimeStamp + _arrow + _subRipTimeStamp);

  if (!format.hasMatch(line)) {
    return null;
  }

  final List<String> times = line.split(_arrow);

  final Duration? start = parseSubRipTimestamp(times[0]);
  final Duration? end = parseSubRipTimestamp(times[1]);

  if (start == null || end == null) {
    return null;
  }

  return [start, end];
}

String extractTextFromHtml(String htmlString) {
  final RegExp exp = RegExp(_html, multiLine: true, caseSensitive: true);
  return htmlString.replaceAll(exp, '');
}

const String _subRipTimeStamp = r'\d\d:\d\d:\d\d,\d\d\d';

const String _arrow = r' --> ';

const String _webVTTSegment = r'(?:(.+?)(?:\r\n?|\n))?'
    r'(\d{2,}:)?(\d{2}:)(\d{2})\.(\d+)\s+-->\s+(\d{2,}:)?(\d{2}:)(\d{2})\.(\d+)\s*?([^\r\n]*?)(?:\r\n?|\n)'
    r'(?<content>[^\0]*?)'
    r'(?=(?:\r\n?|\n)*?(?:$|(?:.+?(?:\r\n?|\n))?.*?-->))';

const String _html = r'<[^>]*>|&[^;]+;';


example/lib/main.dart
// import 'dart:convert' show utf8;

import 'package:chewie/chewie.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_subtitle/flutter_subtitle.dart' hide Subtitle;
import 'package:video_player/video_player.dart';
// import 'package:http/http.dart' as http;

void main() => runApp(const VideoApp());

class VideoApp extends StatefulWidget {
  const VideoApp({Key? key}) : super(key: key);

  @override
  State createState() => _VideoAppState();
}

class _VideoAppState extends State<VideoApp> {
  ChewieController? _chewieController;
  late SubtitleController _subtitleController;

  @override
  void initState() {
    getWebVtt();
    super.initState();
  }

  Future<String> loadAsset() async {
    return await rootBundle.loadString('assets/webvtt.vtt');
  }

  getWebVtt() async {
    // https://cc.zorores.com/20/2e/202eaab6dff289a5976399077449654e/eng-2.vtt
    // https://ohplayer.netlify.app/%E5%90%9B%E3%81%AE%E5%90%8D%E3%81%AF.srt

    // final body = utf8.decode((await http.get(Uri.parse(
    //         'https://cc.zorores.com/20/2e/202eaab6dff289a5976399077449654e/eng-2.vtt')))
    //     .bodyBytes);

    final body = await loadAsset();

    _subtitleController =
        SubtitleController.string(body, format: SubtitleFormat.webvtt);
    final controller = VideoPlayerController.network(
      'https://ohplayer.netlify.app/%E5%90%9B%E3%81%AE%E5%90%8D%E3%81%AF.mp4',
    );

    await controller.initialize();

    controller.addListener(() {
      setState(() {});
    });

    _chewieController = ChewieController(
      videoPlayerController: controller,
      autoPlay: true,
      looping: true,
      subtitleBuilder: (context, subtitle) {
        return IgnorePointer(
          child: SubtitleView(
            text: subtitle,
            subtitleStyle: SubtitleStyle(
              fontSize: _chewieController!.isFullScreen ? 20 : 16,
            ),
          ),
        );
      },
    );

    //// Chewie Subtitle
    _chewieController!.setSubtitle(
      _subtitleController.subtitles
          .map(
            (e) => Subtitle(
              index: e.number,
              start: Duration(milliseconds: e.start),
              end: Duration(milliseconds: e.end),
              text: e.text,
            ),
          )
          .toList(),
    );

    //// Caption
    // final webVTTCaptionFile = _subtitleController.subtitles
    //     .map((e) => Caption(
    //           number: e.number,
    //           start: Duration(milliseconds: e.start),
    //           end: Duration(milliseconds: e.end),
    //           text: e.text,
    //         ))
    //     .toList();

    // controller.setClosedCaptionFile(
    //     Future.value(MyWebVTTCaptionFile(webVTTCaptionFile)));

    controller.setClosedCaptionFile(Future.value(WebVTTCaptionFile(body)));
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Video Demo',
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.start,
          children: [
            const SizedBox(height: 200),
            _chewieController != null
                ? AspectRatio(
                    aspectRatio: 16 / 9,
                    child: Chewie(controller: _chewieController!),
                  )
                : const CircularProgressIndicator(),
            const SizedBox(height: 20),
            const Text('SubtitleControllView'),
            if (_chewieController != null)
              SubtitleControllView(
                subtitleController: _subtitleController,
                inMilliseconds: _chewieController!
                    .videoPlayerController.value.position.inMilliseconds,
              ),
            const SizedBox(height: 20),
            const Text('ClosedCaption'),
            if (_chewieController != null)
              ClosedCaption(
                text:
                    _chewieController!.videoPlayerController.value.caption.text,
                textStyle: const TextStyle(
                  fontSize: 16,
                  color: Colors.white,
                ),
              ),
          ],
        ),
      ),
    );
  }

  @override
  void dispose() {
    super.dispose();
    _chewieController?.dispose();
  }
}

class MyWebVTTCaptionFile extends ClosedCaptionFile {
  MyWebVTTCaptionFile(this.captions);

  @override
  List<Caption> captions = [];
}
